extensions/eclipse-che-theia-testing-service-ext/src/browser/languages-test-main.ts

```diff
@@ -0,0 +1,170 @@
+/********************************************************************************
+ * Copyright (C) 2020 RedHat and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject } from 'inversify';
+import { PluginInfo, LanguagesExt, MAIN_RPC_CONTEXT } from '@theia/plugin-ext/lib/common/plugin-api-rpc';
+import { SerializedDocumentFilter } from '@theia/plugin-ext/lib/common/plugin-api-rpc-model';
+import { LanguagesMainImpl } from '@theia/plugin-ext/lib/main/browser/languages-main';
+import * as theia from '@theia/plugin';
+import { PluginHandleRegistry } from './plugin-handle-registry';
+import { RPCProtocol } from '@theia/plugin-ext/lib/common/rpc-protocol';
+
+export type LanguageServerAction = string;
+export type LanguageServerActions =
+    'completion' |
+    'definition' |
+    'declaration' |
+    'signatureHelp' |
+    'implementation' |
+    'typeDefinition' |
+    'hover' |
+    'documentHighlight' |
+    'workspaceSymbols' |
+    'documentFormattingEdits' |
+    'documentRangeFormattingEdits' |
+    'onTypeFormattingEdits' |
+    'documentLinks' |
+    'codeActions' |
+    'codeLenses' |
+    'references' |
+    'symbols' |
+    'documentColors' |
+    'foldingRange' |
+    'renameEdits';
+
+@injectable()
+export class LanguagesMainTestImpl extends LanguagesMainImpl {
```

Comment:
There really is no reason to extend LanguagesMainImpl. Implementation inheritance is that big stick of OO and should only be used if appropriate (https://en.wikipedia.org/wiki/Fragile_base_class). You can just implement LanguagesMain and delegate to a wrapped instance of LanguagesMainImpl, no?

Reply:
yes agree, it looks quite hard to maintain.
Also it looks like a proxy object